src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -3739,6 +3739,15 @@ const char* InstanceKlass::internal_name() const {
 void InstanceKlass::print_class_load_logging(ClassLoaderData* loader_data,
                                              const ModuleEntry* module_entry,
                                              const ClassFileStream* cfs) const {
+  if (TraceClassLoadingCause != nullptr &&
+      (strcmp(TraceClassLoadingCause, "*") == 0 || strstr(external_name(), TraceClassLoadingCause) != nullptr))
+  {
+    stringStream st;
+    st.print_cr("Loading %s", external_name());
+    JavaThread::current()->print_stack_on(&st);
+    tty->print_raw(st.as_string());
```

Comment:
I assume you mean make the feature require `-Xlog:class+load` in addition? I had that at first but it adds a lot of noise when one is only interested in certain classes as it now shows a UL line for every class loaded, including those not matched by the filter.
I think this is such niche functionality, it's better to keep it separate from `-Xlog:class+load`.

Reply:
I prefer to see this as an extension of UL rather than a stand-alone tracing utility as we have been trying to get rid of the latter and focus on UL. By using UL you also get benefits like easy redirection to a file etc, rather than dumping to tty.

But I understand about the "noise" aspect. Need to think more on this ...

Reply:
I agree that this should be integrated with UL because we tried to remove all of the TraceXXX Flags but it would be good to be able to specify a class to log which UL doesn't have the ability to do.
If you changed tty->print_raw() to log_info(class, load, stacktrace)("%s", st.as_string()); does it put the decorators on each line?

So one thing you could do is add an option like LogClassLoadingCauseFor=x to match with X.  Then when specified with -Xlog:class+load+cause, that will log the right thing.

The code needs a ResourceMark to use external_name(), and you should put this in a static function just above print_class_load_logging() since it doesn't really do the same thing as the rest of print_class_load_logging() but call it from there.   Then if you want to extend it to print native frames, it would be contained.

Also, this is always JavaThread::current() because all the callers have the current JavaThread.  No need for the null check or the cast.

Reply:
As David said, the advantage of using logging is for output file redirection and avoidance of more tty in the code.

Reply:
More.  You could even do -Xlog:class+load+cause+native to get the native stack.  -Xlog:class+load+cause* to get native and Java stacks, and configure it to be compatible with logging rather than these character sequences you have above.

Reply:
`external_name()` needs a `ResourceMark`.

Reply:
These are all good suggestions and feedback. I'll need some help wrt using UL macros. Firstly, how do I expand upon:
```
  if (!log_is_enabled(Info, class, load)) {
    return;
  }
```
to also support `-Xlog:class+load+cause`, `-Xlog:class+load+cause+native` and `-Xlog:class+load+cause*`?

Reply:
```
if (log_is_enabled(Info, class, load, cause)) {
  doStuff();
}
```
HTH

Reply:
Are you saying `log_is_enabled(Info, class, load, cause)` implies `log_is_enabled(Info, class, load)`?
Or do I need:
```
  if (!log_is_enabled(Info, class, load) && !log_is_enabled(Info, class, load, cause)) {
    return;
  }
```
to extend the check on [line 3793](https://github.com/openjdk/jdk/blob/e8ff74c7e84ec2440a51fee1b4c45e87332807a0/src/hotspot/share/oops/instanceKlass.cpp#L3793)?

I don't understand the relationship between `class+load+cause` and `class+load` and the comment on `log_is_enabled` doesn't help much.

Reply:
Usually, the user would need to specify both. Otherwise just having the cause without the `class+load` would be rather useless

```
java -Xlog:class+load  -Xlog:class+load+cause 
```

Which can be combined like:

```
java -Xlog:class+load,class+load+cause 
```

or 

```
java -Xlog:class+load*=info
```

(To see usages of `-Xlog`, you can grep for `[-]Xlog:` in the *.java files under test/hotspot/jtreg)


In terms of implementing this PR, I would suggest handling the class+load and class+load+cause indecently. Something like


```
void InstanceKlass::print_class_load_logging(ClassLoaderData* loader_data,
                                             const ModuleEntry* module_entry,
                                             const ClassFileStream* cfs) const {
  if (ClassListWriter::is_enabled()) {
    ClassListWriter::write(this, cfs);
  }

  log_class_load(...);  // handle -Xlog:class+load 
                        // (the rest of the old print_class_load_logging() function)
  log_class_load_cause(...);  // handle -Xlog:class+load
}
```

The output would be something like

```
[0.014s][info][class,load] java.lang.Thread$State source: shared objects file
[0.014s][info][class,load,cause] 	at jdk.internal.misc.VM.toThreadState(java.base/VM.java:336)
[0.014s][info][class,load,cause] 	at java.lang.Thread.threadState(java.base/Thread.java:2731)
[0.014s][info][class,load,cause] 	at java.lang.Thread.isTerminated(java.base/Thread.java:2738)
[0.014s][info][class,load,cause]	at java.lang.Thread.getThreadGroup(java.base/Thread.java:1957)
...
```

This way, you don't need to repeat the class name, which is already printed by `-Xlog:class+load`

Reply:
> Also, this is always JavaThread::current() because all the callers have the current JavaThread. No need for the null check or the cast.

Does this mean this existing type check and cast is unnecessary: https://github.com/openjdk/jdk/blob/e8ff74c7e84ec2440a51fee1b4c45e87332807a0/src/hotspot/share/oops/instanceKlass.cpp#L3821-L3823

Reply:
Yes. I originally raised the possibility of the current thread not being set yet, but then when I checked it would always be set. But you chose to to add in the check "just in case".

Reply:
> This way, you don't need to repeat the class name, which is already printed by -Xlog:class+load

No this is bad form IMO. It only works if the person enabling logging knows that these two have to be specified together - and there is nothing that actually documents what the different log tags do.

Using multi-level tags is tricky to get right. If you assume they are dependent then specifying one without the other leads to strange looking output because bits are missing. But if you specify both, or use wildcards, you get repetition. This problem arises when you try to use a tag as a "level of detail" rather than as a classification.